### PR TITLE
jcroteau/APPEALS-23894 - Close VeteranRecordRequest tasks for Veterans Readiness and Employment (VRE) business line

### PR DESCRIPTION
**Pull request by jcroteau**
Resolves https://jira.devops.va.gov/browse/APPEALS-23894

UAT PR: https://github.com/department-of-veterans-affairs/caseflow/pull/18983

# Description

> From related SNOW ticket: [APPEALS-4106 - Manually Complete VeteranRecordRequest INC22451913](https://jira.devops.va.gov/browse/APPEALS-4106):
> 
> >Caseflow is automatically creating VeteranRecordRequest tasks for the business line VRE. This is incorrect as VRE now uses VBMS and we don't need to get the records manually. There are currently 136 active VRE record request tasks that should be set to completed or cancelled. The tasks can be found in the database with the query `SELECT * from tasks WHERE type='VeteranRecordRequest' AND status in ('assigned', 'on_hold', 'in_progress') AND assigned_to_id=220;`
> 

- **Note:** As of this writing, the root cause of these errant tasks being created has not yet been addressed in the application code. See related [Slack thread](https://benefits-int-delivery.slack.com/archives/C03ML6AG7V3/p1687816546000239?thread_ts=1687816524.487899&cid=C03ML6AG7V3).

> 
> >As a tier 4 developer I need to research and document the manual remediation steps to cancel all open Veteran Record Requests assigned to the Vocational Readiness and Education business line and its users so that Appeals can no longer be blocked by these tasks.
>
> ## Acceptance Criteria
> - Remediation steps are identified and documented to
>     - [x] Find all open (IN\_PROGRESS, ON\_HOLD, ASSIGNED) Veteran Record Requests tasks assigned to
>         - The Vocational Readiness and Education business (id 220 in prod)
>         - Users belonging to the Vocational Readiness and Education org
>     - [x] Add logging to report metrics
>         - How many tasks are we attempting close
>         - How many were closed at the end of the script
>         - How long each one took
>         - List of tasks and errors
>     - [x] Loop through all open tasks identified and set status to cancelled

- Introduces rake task `remediations:cancel_vrr_tasks_open_for_vre`, which identifies any currently open `VeteranRecordReqeust` Tasks that are assigned to the **"Veteran Readiness and Employment"** `BusinessLine` and cancels them and their open descendant tasks.

## Testing Plan
Jira Test Issue: https://jira.devops.va.gov/browse/APPEALS-25089

#### Setup
- Access the Rails console:
```bash
sudo su - 
source /opt/caseflow-certification/caseflow-certification_env.sh
cd /opt/caseflow-certification/src
bin/rails console
```
- [x] In the Rails console, find some test Veterans and note their `file_number` s
```ruby
file_number_1, file_number_2 = Veteran.order(created_at: :asc).last(2).map(&:file_number)
```

### Find initial count of open and cancelled VRR tasks for VRE
- [x] From the Rails console, run the following query and note the results:
```ruby
vre_business_line =  
  BusinessLine.where(name: Constants::BENEFIT_TYPES["voc_rehab"]);
  
initial_count_of_open_vrr_tasks = 
  VeteranRecordRequest.open.where(assigned_to: vre_business_line).count

initial_count_of_cancelled_vrr_tasks = 
  VeteranRecordRequest.cancelled.where(assigned_to: vre_business_line).count
```

### Create new VRR tasks for VRE

**Establish Appeals w/ issues of Benefit type "Veterans Readiness and Employment" to create new VRR tasks for VRE:**

- [ ] Switch to a user w/ permissions to intake appeals (`BVAISHAW` in dev)

#### Establish Appeal for test Veteran 1
- [ ] Visit `/intake` page
- [ ] Select `Decision Review Request: Board Appeal (Notice of Disagreement) — VA Form 10182`
- [ ] Enter Veteran's `file_number` (noted above)
- [ ] Fill out and submit Review form:
    - **What is the Receipt Date of this form?** Yesterday
    - **Was this form submitted through VA.gov?** No
    - **Which review option did the Veteran request?** Direct Review
    - **Is the claimant someone other than the Veteran?** No
    - **Did the Veteran check the "OPT-IN from SOC/SSOC" box on the form?** N/A
- [ ] Add an issue w/ Benefit type `Veteran Readiness and Employment`
    - **Benefit type** Veterans Readiness and Employment
    - **Issue category** Additional Training
    - **Decision date** Yesterday
    - **Issue description** foobar
- [ ] Add an issue w/ Benefit type `Education`
    - **Benefit type** Education
    - **Issue category** Accrued
    - **Decision date** Yesterday
    - **Issue description** foobar
- [ ] Click `Establish appeal`
- [ ] Confirm that you get an `Intake completed` confirmation
- [ ] From the Rails console, confirm that a `VeteranRecordRequest` task was created for the appeal and assigned to the VRE business line:
```ruby
appeal_1 = Appeal.where(veteran_file_number: file_number_1).order(:created_at).last;
appeal_1.uuid
appeal_1.request_issues.first.benefit_type # => "voc_rehab"
vrr_task_1 = VeteranRecordRequest.where(appeal: appeal).first
vrr_task_1.assigned_to.name # => "Veterans Readiness and Employment"
```
- [ ] Note the appeal `uuid`
#### Establish Appeal for test Veteran 2
- [ ] Repeat the Intake steps above for the second test Veteran
line:
```ruby
appeal_2 = Appeal.where(veteran_file_number: file_number_2).order(:created_at).last;
appeal_2.uuid
appeal_2.request_issues.first.benefit_type # => "voc_rehab"
vrr_task_2 = VeteranRecordRequest.where(appeal: appeal).first
vrr_task_2.assigned_to.name # => "Veterans Readiness and Employment"
```
### Find updated count of existing, open VRR tasks for VRE
- [ ] From the Rails console, run the following query and note the result:
```ruby
vre_business_line =  
  BusinessLine.where(name: Constants::BENEFIT_TYPES["voc_rehab"])

updated_count_of_open_vrr_tasks =
  VeteranRecordRequest.open.where(assigned_to: vre_business_line).count
```
- [ ] The count should now be `2` **more** than the initial count (`initial_count_of_open_vrr_tasks`).
```ruby
updated_count_of_open_vrr_tasks == (initial_count_of_open_vrr_tasks + 2)

#=> true
```
### Run rake task to cancel existing open VRR tasks for VRE
- [x] cd into the Caseflow app root directory:
    - Local dev: `cd /dev/appeals/caseflow`
    - UAT:  `cd /opt/caseflow-certification/src`
- [x] Run rake task:
```bash
bin/rake remediations:cancel_vrr_tasks_open_for_vre
```
### Find final count of existing open and cancelled VRR tasks for VRE
- [x] From the Rails console, run the following query and note the result:
```ruby
vre_business_line =  
  BusinessLine.where(name: Constants::BENEFIT_TYPES["voc_rehab"])
  
final_count_of_open_vrr_tasks = 
  VeteranRecordRequest.open.where(assigned_to: vre_business_line).count

final_count_of_cancelled_vrr_tasks = 
  VeteranRecordRequest.cancelled.where(assigned_to: vre_business_line).count
```
- [x] The final count of open VRR tasks for VRE should now be `0`, and the final count of cancelled VRR tasks for VRE should be `2` more than the initial count:
```ruby
final_count_of_open_vrr_tasks == 0

# => true

final_count_of_cancelled_vrr_tasks == (initial_count_of_cancelled_vrr_tasks + 2)

# => true
```

---

- [ ] For feature branches merging into master: Was this deployed to UAT?

# Best practices
## Code Documentation Updates
- [x] Add or update code comments at the top of the class, module, and/or component.

### Test Coverage
Did you include any test coverage for your code? Check below:
- [x] RSpec
- [ ] Jest
- [ ] Other

### Code Climate
Your code does not add any new code climate offenses? If so why?
- [ ] No new code climate issues added

### Monitoring
`N/A`

### Logging
- [x] Are logs being produced at appropriate log levels (debug, info, warn, error, fatal)?
- [x] Are logs structured (e.g., using log tags) for easier querying and analysis?
- [ ] Are sensitive data (e.g., passwords, tokens) redacted or omitted from logs?
- [ ] Is log retention and rotation configured correctly?
- [x] Are logs being forwarded to a centralized logging system if needed?

### Auditing
- [x] Are user actions being logged for audit purposes?
- [x] Are changes to critical data being tracked ?
- [ ] Are logs being securely stored and protected from tampering or exposing protected data?

### Error Handling
- [x] Are errors being caught and handled gracefully?
- [x] Are appropriate error messages being displayed to users?
- [ ] Are critical errors being reported to an error tracking system (e.g., Sentry, ELK)?
- [ ] Are unhandled exceptions being caught at the application level ?

### Exception Handling
- [x] Are custom exceptions defined and used where appropriate?
- [x] Is exception handling consistent throughout the codebase?
- [x] Are exceptions logged with relevant context and stack trace information?
- [x] Are exceptions being grouped and categorized for easier analysis and resolution?

### CODE REVIEW

File: lib/helpers/cancel_tasks_and_descendants.rb [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  def initialize(task_relation)
+    @task_relation = task_relation
+    @request_id = SecureRandom.uuid
+  end
+
+  def call
+    RequestStore[:current_user] = User.system_user
+
+    log_time_elapsed { log_task_count_before_and_after { cancel_tasks } }
+  end
```
Comment (jcroteau) [inline lib/helpers/cancel_tasks_and_descendants.rb:27]: 📝 The block-iness of this pattern might seem a little bit too "clever" upon first read, but my goal was to keep the logging as visually and logically separated from the business logic as possible. 

Here is an example sequence of log entries:
```
[CancelTasksAndDescendants][47703dac-6828-4158-9f76-ded6c7540b16] Total tasks for cancellation: 9
[CancelTasksAndDescendants][47703dac-6828-4158-9f76-ded6c7540b16] Task ids [1, 2, 3] cancelled successfully
[CancelTasksAndDescendants][47703dac-6828-4158-9f76-ded6c7540b16] Task ids [4, 5, 6] cancelled successfully
[CancelTasksAndDescendants][47703dac-6828-4158-9f76-ded6c7540b16] Task ids [7, 8] not cancelled due to error - <error message>
[CancelTasksAndDescendants][47703dac-6828-4158-9f76-ded6c7540b16] Task ids [9] not cancelled due to error - <error message>
[CancelTasksAndDescendants][47703dac-6828-4158-9f76-ded6c7540b16] Tasks cancelled successfully: 6
[CancelTasksAndDescendants][47703dac-6828-4158-9f76-ded6c7540b16] Elapsed time (sec): 0.005013999994844198
```

File: lib/tasks/remediations/cancel_vrr_tasks_open_for_vre.rake [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+require_relative "../../../lib/helpers/cancel_tasks_and_descendants"
+
+namespace :remediations do
+  desc "Cancel VeteranRecordRequest tasks that are both open and assigned to " \
+       "the 'Veterans Readiness and Employment' business line"
+  task :cancel_vrr_tasks_open_for_vre => [:environment] do
+    CancelTasksAndDescendants.call(
+      VeteranRecordRequestsOpenForVREQuery.call
+    )
```
Comment (jcroteau) [inline lib/tasks/remediations/cancel_vrr_tasks_open_for_vre.rake:11]: 📝 **Note to Reviewers:**

My rationale for injecting the Task relation (`VeteranRecordRequestsOpenForVREQuery.call`) was two-fold:
(1) It makes the testing easier (as dependency injection often does)
(2) It makes `CancelTasksAndDescendants` more re-usable. I figured cancelling tasks is probably a fairly common thing to do, so we could re-use this class with other task relations to handle different remediation scenarios.

File: lib/helpers/cancel_tasks_and_descendants.rb [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+  def call
+    RequestStore[:current_user] = User.system_user
+
+    with_paper_trail_options do
+      log_time_elapsed { log_task_count_before_and_after { cancel_tasks } }
+      print_logs_to_stdout
+    end
+  end
```
Comment (jcroteau) [inline lib/helpers/cancel_tasks_and_descendants.rb:31]: 📝 **Note to Reviewers:** 

The liberal use of blocks might seem a little bit too "clever" upon first read, but my goal in this was to keep the logging as visually and logically distinct from the business logic as much as possible.

In theory, if we decide to dramatically change the format or destination of the logs (e.g. AWS S3), the changes should be isolated to the dedicated logging methods, away from the business logic of finding and cancelling tasks.

If anyone can think of a more elegant way to do this, I'm all ears 😄 

My aim is to eventually arrive at a logging pattern that could be easily adopted in our other Tier 4 scripts.

Here is an example sequence of log entries:
```
[CancelTasksAndDescendants][47703dac-6828-4158-9f76-ded6c7540b16] Total tasks for cancellation: 9
[CancelTasksAndDescendants][47703dac-6828-4158-9f76-ded6c7540b16] Task ids [1, 2, 3] cancelled successfully
[CancelTasksAndDescendants][47703dac-6828-4158-9f76-ded6c7540b16] Task ids [4, 5, 6] cancelled successfully
[CancelTasksAndDescendants][47703dac-6828-4158-9f76-ded6c7540b16] Task ids [7, 8] not cancelled due to error - <error message>
[CancelTasksAndDescendants][47703dac-6828-4158-9f76-ded6c7540b16] Task ids [9] not cancelled due to error - <error message>
[CancelTasksAndDescendants][47703dac-6828-4158-9f76-ded6c7540b16] Tasks cancelled successfully: 6
[CancelTasksAndDescendants][47703dac-6828-4158-9f76-ded6c7540b16] Elapsed time (sec): 0.005013999994844198
```

File: lib/helpers/cancel_tasks_and_descendants.rb [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  def log_errored(task, error)
+    task_ids = cancellable_descendants_for(task).pluck(:id)
+    log("Task ids #{task_ids} not cancelled due to error - #{error}",
+        level: :error)
+  end
+
+  def cancellable_descendants_for(task)
+    # Note: The result of `Task #descendants` also includes the instance itself
+    Task.open.where(id: task.descendants)
+  end
```
Comment (jcroteau) [inline lib/helpers/cancel_tasks_and_descendants.rb:68]: 📝 **Note to Reviewers:**

Mirrors the relation in [`Task #cancel_task_and_child_subtasks`](https://github.com/department-of-veterans-affairs/caseflow/blob/d62a32f339251d8c8f2de50d70cba294d2020903/app/models/task.rb#L726-L730).

File: lib/helpers/cancel_tasks_and_descendants.rb [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+class CancelTasksAndDescendants
+  LOG_TAG = "CancelTasksAndDescendants"
+
+  # Cancels all tasks and descendant tasks for given Task relation
+  #
+  # @param task_relation [ActiveRecord::Relation] tasks to be cancelled
+  # @return [NilClass]
+  def self.call(task_relation = Task.none)
+    new(task_relation).__send__(:call)
```
Comment (raymond-hughes) [inline lib/helpers/cancel_tasks_and_descendants.rb:13]: I like this. It allows us to hide `new` while still giving this class control. Great idea. We should use this on our SOP updates.

File: lib/helpers/cancel_tasks_and_descendants.rb (hunk 3 above)
Comment (raymond-hughes) [reply lib/helpers/cancel_tasks_and_descendants.rb:31]: Great explanation. It is a good separation of concerns and I like this.

File: lib/helpers/cancel_tasks_and_descendants.rb [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  end
+
+  def append_to_logs_for_stdout(message)
+    @logs << "[#{LOG_TAG}] [#{@request_id}] #{message}"
+  end
+
+  def print_logs_to_stdout
+    puts @logs
+  end
+end
```
Comment (raymond-hughes) [inline lib/helpers/cancel_tasks_and_descendants.rb:117]: I am wondering if we need to pull the logging out into a separate PR? Do you think we can add this as a mixin? I see your reasoning and we can discuss further.

Comment (jcroteau) [reply lib/helpers/cancel_tasks_and_descendants.rb:117]: I would like to eventually get us to a place where logging concerns can be abstracted from our Tier 4 scripts, but it might take some experimentation and a few iterations to get there. There is a lot of logging stuff that is specific to this script that would be difficult to cleanly lob off into a module or something.

What I'd really like is for us to have JSON-formatted logs with a uniform schema, that we can then more easily query using Kibana, but I didn't want to venture there quite yet, as some thought will need to be put into the schema. 

Also, I didn't know if the project as-a-whole was trying to move toward a uniform JSON schema for our application logs (a la `logstasher`). It looks like maybe there was some movement in this direction, but perhaps it stalled out.

That might be a good question for you to bring to the TL sync:

- ❓ Is the project trying to move to a uniform logging format (JSON) and schema for easier log analysis. If not, should we?

**Comment by jcroteau** [comment]
test comment

**Comment by jcroteau** [comment]
@jcroteau test comment

**Comment by jcroteau** [comment]
@jcroteau test comment

**Comment by jcroteau** [comment]
@jcroteau test

**Comment by jcroteau** [comment]
@jcroteau test

**Comment by jcroteau** [comment]
test

**Comment by jcroteau** [comment]
test

**Comment by jcroteau** [comment]
test test test

**Comment by jcroteau** [comment]
test

**Review by raymond-hughes — APPROVED** [review]
